tools/core2js.rkt

```diff
@@ -0,0 +1,186 @@
+#lang racket
+
+(require "common.rkt" "fpcore.rkt")
+(provide compile-program)
+
+(define (fix-name name)
+  (string-join
+    (for/list ([char (~a name)])
+      (if (regexp-match #rx"[a-zA-Z0-9_]" (string char))
+          (string char)
+          (format "_~a_" (char->integer char))))
+    ""))
+
+(define/match (constant->js c)
+  [('E) "math.E"]
+  [('LOG2E) "math.LOG2E"]
+  [('LOG10E) "math.LOG10E"]
+  [('LN2) "math.LN2"]
+  [('LN10) "math.LN10"]
+  [('PI) "math.pi"]
+  [('PI_2) "(math.pi/2)"]
+  [('PI_4) "(math.pi/4)"]
+  [('1_PI) "(1/math.pi)"]
+  [('2_PI) "(2/math.pi)"]
+  [('2_SQRTPI) "(2/math.sqrt(math.pi))"]
+  [('SQRT2) "math.SQRT2"]
+  [('SQRT1_2) "(1/math.SQRT2)"]
+  [('MAXFLOAT) "Number.MAX_VALUE"]
+  [('TRUE) "true"]
+  [('FALSE) "false"]
+  [('INFINITY) "Infinity"]
+  [('NAN) "math.NaN"]
+  [(_) (error 'constant->js "Unsupported constant ~a" c)])
+
+(define/match (operator->js op)
+  [((or '== '+ '- '* '/  '< '> '<= '>=)) (format "(~a ~a ~a)" "~a" op "~a")]
+  [('fabs) "math.abs(~a)"]
+  [('exp) "math.exp(~a)"]
+  [('exp2) "math.pow(2, ~a)"]
+  [('expm1) "math.expm1(~a)"]
+  [('log) "math.log(~a)"]
+  [('log10) "math.log10(~a)"]
+  [('log2) "math.log2(~a)"]
+  [('log1p) "math.log1p(~a)"]
+  [('logb) "math.floor(math.log2(math.abs(~a)))"]
+  [('pow) "math.pow(~a, ~a)"]
+  [('sqrt) "math.sqrt(~a)"]
+  [('cbrt) "math.cbrt(~a)"]
+  [('hypot) "math.hypot(~a, ~a)"]
+  [('sin) "math.sin(~a)"]
+  [('cos) "math.cos(~a)"]
+  [('tan) "math.tan(~a)"]
+  [('asin) "math.asin(~a)"]
+  [('acos) "math.acos(~a)"]
+  [('atan) "math.atan(~a)"]
+  [('atan2) "math.atan2(~a, ~a)"]
+  [('sinh) "math.sinh(~a)"]
+  [('cosh) "math.cosh(~a)"]
+  [('tanh) "math.tanh(~a)"]
+  [('asinh) "math.asinh(~a)"]
+  [('acosh) "math.acosh(~a)"]
+  [('atanh) "math.atanh(~a)"]
+  [('erf) "math.erf(~a)"]
+  [('erfc) "1 - math.erf(~a)"]
+  [('tgamma) "math.gamma(~a)"]
+  [('lgamma) "math.log(math.gamma(~a))"]
+  [('ceil) "math.ceil(~a)"]
+  [('floor) "math.floor(~a)"]
+  [('remainder) "math.mod(~a, ~a)"]
+  [('fmax) "math.max(~a, ~a)"]
+  [('fmin) "math.min(~a, ~a)"]
+  [('fdim) "math.max(0, ~a - ~a)"]
+  [('copysign) "math.abs(~a) * math.sign(~a)"]
+  [('trunc) "math.fix(~a)"]
+  [('round) "math.round(~a)"]
+  [('isinf)  "(math.abs(~a) === Number.POSITIVE_INFINITY)"]
+  [('isnan) "isNaN(~a)"]
+  [(_) (error 'operator->js "Unsupported operation ~a" op)])
+
+(define (application->js type operator args)
+  (if (and (eq? operator '-) (= (length args) 1))
+    (format "(- ~a)" (car args))
+    ((apply (curry format) (cons (operator->js operator) args)))))
```

Comment:
I don't think `curry` is doing anything here. Also, you can replace `(apply f (cons x xs))` with `(apply f x xs)`.